core/arch/arm/plat-stm32mp1/plat_tzc400.c

```diff
@@ -33,91 +114,440 @@ static enum itr_return tzc_it_handler(struct itr_handler *handler __unused)
-	uint32_t filters_mask = TZC_FILTERS_MASK;
-	vaddr_t base = pa;
+	uintptr_t base = tzc_dev->pdata.base;
+	uint32_t regval = 0;
 
-	if (tzc_get_region_config(i, &region_cfg))
-		panic();
+	regval = io_read32(base + BUILD_CONFIG_OFF);
+	tzc_dev->ddata.nb_filters = ((regval >> BUILD_CONFIG_NF_SHIFT) &
+				     BUILD_CONFIG_NF_MASK) + 1;
+	tzc_dev->ddata.nb_regions = ((regval >>	BUILD_CONFIG_NR_SHIFT) &
+				     BUILD_CONFIG_NR_MASK);
 
-	return region_cfg.base == base && region_cfg.top == (base + size - 1) &&
-	       region_cfg.sec_attr == TZC_REGION_S_NONE &&
-	       (region_cfg.ns_device_access & ns_cpu_mask) == ns_cpu_mask &&
-	       region_cfg.filters == filters_mask;
+	DMSG("TZC400 Filters %"PRIu32" Regions %"PRIu32,
+	     tzc_dev->ddata.nb_filters, tzc_dev->ddata.nb_regions);
 }
 
-static bool tzc_region_is_secure(unsigned int i, vaddr_t base, size_t size)
+static void stm32mp_tzc_region0(bool enable)
 {
-	struct tzc_region_config region_cfg = { };
-	uint32_t filters_mask = TZC_FILTERS_MASK;
+	struct tzc_region_config region_cfg_0 = {
+		.base = 0,
+		.top = UINT_MAX,
+		.sec_attr = TZC_REGION_S_NONE,
+		.ns_device_access = 0,
+	};
 
-	if (tzc_get_region_config(i, &region_cfg))
-		panic();
+	if (enable)
+		region_cfg_0.sec_attr = TZC_REGION_S_RDWR;
+
+	tzc_configure_region(0, &region_cfg_0);
+}
+
+static void stm32mp_tzc_reset_region(struct tzc_device *tzc_dev)
+{
+	unsigned int i = 0;
+	const struct tzc_region_config cfg = { .top = 0x00000FFF };
+
+	/* Clean old configuration */
+	for (i = 0; i < tzc_dev->ddata.nb_regions; i++)
+		tzc_configure_region(i + 1, &cfg);
+}
+
+static TEE_Result stm32mp_tzc_region_append(struct tzc_device *tzc_dev,
+					    const struct tzc_region_config
+					    *region_cfg)
+{
+	TEE_Result res = TEE_SUCCESS;
+	unsigned int index = 0;
+	index = tzc_dev->nb_reg_used;
+
+	if (index >= tzc_dev->ddata.nb_regions ||
+	    !core_is_buffer_inside(region_cfg->base,
+				   region_cfg->top + 1 - region_cfg->base,
+				   tzc_dev->pdata.mem_base,
+				   tzc_dev->pdata.mem_size))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = tzc_region_check_overlap(tzc_dev, region_cfg);
+	if (res)
+		return res;
+
+	tzc_dev->reg[tzc_dev->nb_reg_used] = *region_cfg;
+	tzc_dev->nb_reg_used++;
+
+	tzc_configure_region(tzc_dev->nb_reg_used, region_cfg);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result
+exclude_region_from_nsec(const struct tzc_region_config *reg_exclude)
+{
+	struct tzc_region_non_sec *reg = NULL;
+
+	SLIST_FOREACH(reg, &nsec_region_list, link) {
+		if (core_is_buffer_inside(reg_exclude->base,
+					  reg_exclude->top + 1 -
+					  reg_exclude->base,
+					  reg->region.base,
+					  reg->region.top + 1 -
+					  reg->region.base))
+			break;
+	}
+
+	if (!reg)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (reg_exclude->base == reg->region.base &&
+	    reg_exclude->top == reg->region.top) {
+		/* Remove this entry */
+		SLIST_REMOVE(&nsec_region_list, reg, tzc_region_non_sec, link);
```

Comment:
`free(reg)`?

Reply:
I'm changing to SLIST_FOREACH_SAFE above then.